File: 0283-move-zeroes/0283-move-zeroes.java
class Solution {
    public void moveZeroes(int[] nums) {
        int counter = 0;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                nums[counter] = nums[i];
                counter++;
            }
        }

        for (int i = counter; i < nums.length; i++) {
            // counter is the number of non-zeroes in the array
            // and the starting position from which to fill zeroes
            nums[i] = 0;
        }
    }
}

File: 0283-move-zeroes/0283-move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # two pointers approach
        # have a starting index for non-zero numbers
        # fill with zeroes from this pointer to the end

        i = 0
        for num in nums:
            if num != 0:
                nums[i] = num
                i += 1
        
        for j in range(i, len(nums)):
            nums[j] = 0